framework/src/functions/VectorPostprocessorFunction.C

```diff
@@ -40,12 +40,16 @@ VectorPostprocessorFunction::validParams()
 VectorPostprocessorFunction::VectorPostprocessorFunction(const InputParameters & parameters)
   : Function(parameters),
     VectorPostprocessorInterface(this),
-    _component(parameters.get<unsigned int>("component")),
     _argument_column(getVectorPostprocessorValue("vectorpostprocessor_name",
                                                  getParam<std::string>("argument_column"))),
     _value_column(getVectorPostprocessorValue("vectorpostprocessor_name",
-                                              getParam<std::string>("value_column")))
+                                              getParam<std::string>("value_column"))),
+    _has_component(isParamValid("component"))
 {
+  // determine function argument
+  if (_has_component)
+    _component = parameters.get<unsigned int>("component");
+
```

Comment:
_component(isParamValid("component") ? getParam<unsigned int>("component") : NO_COMPONENT),

Reply:
NO_COMPONENT can be a static member variable set to std::numeric_limits<unsigned int>::max();

Reply:
You can use another name besides NO_COMPONENT

Reply:
When I switched to `_component(isParamValid("component") ? getParam("component") : NO_COMPONENT),`

I get the following compiler error:

```
error: no matching function for call to ‘VectorPostprocessorFunction::getParam(const char [10])’
   47 |     _component(isParamValid("component") ? getParam("component") : NO_COMPONENT)
      |                                                                ^
In file included from /home/crswong888/projects/moose/framework/build/header_symlinks/Function.h:12,
                 from /home/crswong888/projects/moose/framework/build/header_symlinks/Axisymmetric2D3DSolutionFunction.h:12,
                 from /home/crswong888/projects/moose/framework/src/functions/Axisymmetric2D3DSolutionFunction.C:11,
                 from /home/crswong888/projects/moose/framework/build/unity_src/functions_Unity.C:2:
/home/crswong888/projects/moose/framework/build/header_symlinks/MooseObject.h:207:1: note: candidate: ‘template<class T> const T& MooseObject::getParam(const string&) const’
  207 | MooseObject::getParam(const std::string & name) const
      | ^~~~~~~~~~~
/home/crswong888/projects/moose/framework/build/header_symlinks/MooseObject.h:207:1: note:   template argument deduction/substitution failed:
In file included from /home/crswong888/projects/moose/framework/build/unity_src/functions_Unity.C:27:
/home/crswong888/projects/moose/framework/src/functions/VectorPostprocessorFunction.C:47:64: note:   couldn’t deduce template parameter ‘T’
   47 |     _component(isParamValid("component") ? getParam("component") : NO_COMPONENT)
      |                                                                ^
```

I also am not sure what you mean by setting `NO_COMPONENT` as a static member variable. Right now, the compiler's complaining that I haven't declare `NO_COMPONENT`, which is true obviously...

Reply:
I really just went about this the exact same way PiecewiseLinearBase.C did:

```
Real
PiecewiseLinearBase::value(Real t, const Point & p) const
{
  const Real x = _has_axis ? p(_axis) : t;
  return _scale_factor * _linear_interp->sample(x);
}
```

Except instead of `_has_axis`, my bool was named `_has_component` to be consistent with the original developer of this code.

Reply:
You need to template argument on getParam: `getParam<unsigned int>`

Reply:
https://en.cppreference.com/w/cpp/language/static

Reply:
Yes, I figured it out thank you!